src/integration/data/vectors/upsertAndUpdate.test.ts

```diff
@@ -76,3 +84,93 @@ describe('upsert and update to serverless index', () => {
     updateSpy.mockRestore();
   });
 });
+
+// Retry logic tests
+describe('Testing retry logic on Upsert operation, as run on a mock, in-memory http server', () => {
+  const recordsToUpsert = generateRecords({
+    dimension: 2,
+    quantity: 1,
+    withSparseValues: false,
+    withMetadata: true,
+  });
+
+  let server: http.Server; // Note: server cannot be something like an express server due to conflicts w/edge runtime
```

Comment:
Mock server set up per test in order to 'mock' 5xx errors in a more realistic way than we do in our unit tests. Need to start server anew each test in order to avoid messy state collisions